Sources/Algorithms/EndsWith.swift

```diff
@@ -0,0 +1,82 @@
+
+//===----------------------------------------------------------------------===//
+// EndsWith
+//===----------------------------------------------------------------------===//
+
+extension BidirectionalCollection where Element: Equatable {
+
+
+  /// Returns a Boolean value indicating whether the final elements of the
+  /// collection are the same as the elements in another collection.
+  ///
+  /// This example tests whether one countable range ends with the elements
+  /// of another countable range.
+  ///
+  ///     let a = 8...10
+  ///     let b = 1...10
+  ///
+  ///     print(b.ends(with: a))
+  ///     // Prints "true"
+  ///
+  /// Passing a collection with no elements or an empty collection as
+  /// `possibleSuffix` always results in `true`.
+  ///
+  ///     print(b.ends(with: []))
+  ///     // Prints "true"
+  ///
+  /// - Parameter possibleSuffix: A collection to compare to this collection.
+  /// - Returns: `true` if the initial elements of the collection are the same as
+  ///   the elements of `possibleSuffix`; otherwise, `false`. If
+  ///   `possibleSuffix` has no elements, the return value is `true`.
+  ///
+  /// - Complexity: O(*m*), where *m* is the lesser of the length of the
+  ///   collection and the length of `possibleSuffix`.
+  @inlinable
+  public func ends<PossibleSuffix: BidirectionalCollection>(
+    with possibleSuffix: PossibleSuffix
+  ) -> Bool where PossibleSuffix.Element == Element {
+    return self.ends(with: possibleSuffix, by: ==)
+  }
+
+  /// Returns a Boolean value indicating whether the final elements of the
+  /// collection are equivalent to the elements in another collection, using
+  /// the given predicate as the equivalence test.
+  ///
+  /// The predicate must be an *equivalence relation* over the elements. That
+  /// is, for any elements `a`, `b`, and `c`, the following conditions must
+  /// hold:
+  ///
+  /// - `areEquivalent(a, a)` is always `true`. (Reflexivity)
+  /// - `areEquivalent(a, b)` implies `areEquivalent(b, a)`. (Symmetry)
+  /// - If `areEquivalent(a, b)` and `areEquivalent(b, c)` are both `true`, then
+  ///   `areEquivalent(a, c)` is also `true`. (Transitivity)
+  ///
+  /// - Parameters:
+  ///   - possibleSuffix: A collection to compare to this collection.
+  ///   - areEquivalent: A predicate that returns `true` if its two arguments
+  ///     are equivalent; otherwise, `false`.
+  /// - Returns: `true` if the initial elements of the collection are equivalent
+  ///   to the elements of `possibleSuffix`; otherwise, `false`. If
+  ///   `possibleSuffix` has no elements, the return value is `true`.
+  ///
+  /// - Complexity: O(*m*), where *m* is the lesser of the length of the
+  ///   collection and the length of `possibleSuffix`.
+  @inlinable
+  public func ends<PossibleSuffix: BidirectionalCollection>(
+    with possibleSuffix: PossibleSuffix,
+    by areEquivalent: (Element, PossibleSuffix.Element) throws -> Bool
```

Comment:
Since equivalence is dictated by `areEquivalent` does this need to be constrained to element is `Equatable`(`BidirectionalCollection where Element: Equatable`)?

Reply:
Nice catch! I moved this overload into its own unconstrained extension.

https://github.com/apple/swift-algorithms/compare/17dfa21b45e1d3fdfd31e9f5e3a2c361f4a7cef6..f49094216b7f225dda7ee54ec7ad6c5801dfb75c